File: LeetCode/52N-QueensII/N-QueensII.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        # Backtracking -> Time:O(n^n), Space:O(n)
        col = set()
        pos_diag = set()
        neg_diag = set()
        ans = 0
        def backtrack(r):
            if r == n:
                nonlocal ans
                ans += 1
                return
            for c in range(n):
                if c not in col and (r + c) not in pos_diag and (r - c) not in neg_diag:
                    col.add(c)
                    pos_diag.add(r+c)
                    neg_diag.add(r-c)
                    backtrack(r+1)
                    col.remove(c)
                    pos_diag.remove(r+c)
                    neg_diag.remove(r-c)
        backtrack(0)
        return ans

File: LeetCode/52N-QueensII/N-QueensII.cpp
class Solution {
public:
    void backtrack(int &ans, set<int> &posDiag, set<int> &negDiag, set<int> &cols, int n, int r) {
        if (r == n) {
            ans++;
            return;
        }
        for (int c = 0; c < n; c++) {
            if (posDiag.find(r+c) == posDiag.end() && negDiag.find(r-c) == negDiag.end() && cols.find(c) == cols.end()) {
                cols.insert(c);
                posDiag.insert(r+c);
                negDiag.insert(r-c);
                backtrack(ans, posDiag, negDiag, cols, n, r+1);
                cols.erase(c);
                posDiag.erase(r+c);
                negDiag.erase(r-c);
            }
        }
    }
    int totalNQueens(int n) {
        // Backtracking -> Time:O(n^n), Space:O(n)
        set<int> posDiag;
        set<int> negDiag;
        set<int> cols;
        int ans = 0;
        backtrack(ans, posDiag, negDiag, cols, n, 0);
        return ans;
    }
};
